bundle/config/validate/folder_permissions.go

```diff
@@ -0,0 +1,154 @@
+package validate
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/bundle/config/resources"
+	"github.com/databricks/cli/bundle/permissions"
+	"github.com/databricks/cli/libs/diag"
+	"github.com/databricks/databricks-sdk-go/service/workspace"
+	"golang.org/x/sync/errgroup"
+)
+
+type folderPermissions struct {
+}
+
+// Apply implements bundle.ReadOnlyMutator.
+func (f *folderPermissions) Apply(ctx context.Context, b bundle.ReadOnlyBundle) diag.Diagnostics {
+	if len(b.Config().Permissions) == 0 {
+		return nil
+	}
+
+	paths := []string{b.Config().Workspace.RootPath}
+
+	if !strings.HasPrefix(b.Config().Workspace.ArtifactPath, b.Config().Workspace.RootPath) {
+		paths = append(paths, b.Config().Workspace.ArtifactPath)
+	}
+
+	if !strings.HasPrefix(b.Config().Workspace.FilePath, b.Config().Workspace.RootPath) {
+		paths = append(paths, b.Config().Workspace.FilePath)
+	}
+
+	if !strings.HasPrefix(b.Config().Workspace.StatePath, b.Config().Workspace.RootPath) {
+		paths = append(paths, b.Config().Workspace.StatePath)
+	}
+
+	if !strings.HasPrefix(b.Config().Workspace.ResourcePath, b.Config().Workspace.RootPath) {
+		paths = append(paths, b.Config().Workspace.ResourcePath)
+	}
+
+	var diags diag.Diagnostics
+	errGrp, errCtx := errgroup.WithContext(ctx)
+	for _, path := range paths {
+		p := path
+		errGrp.Go(func() error {
+			diags = diags.Extend(checkFolderPermission(errCtx, b, p))
+			return nil
+		})
+	}
+
+	if err := errGrp.Wait(); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+func checkFolderPermission(ctx context.Context, b bundle.ReadOnlyBundle, folderPath string) diag.Diagnostics {
+	var diags diag.Diagnostics
+	w := b.WorkspaceClient().Workspace
+	obj, err := w.GetStatusByPath(ctx, folderPath)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	objPermissions, err := w.GetPermissions(ctx, workspace.GetWorkspaceObjectPermissionsRequest{
+		WorkspaceObjectId:   fmt.Sprint(obj.ObjectId),
+		WorkspaceObjectType: "directories",
+	})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if len(objPermissions.AccessControlList) != len(b.Config().Permissions) {
+		diags = diags.Append(diag.Diagnostic{
+			Severity: diag.Warning,
+			Summary:  "permissions count mismatch",
+			Detail:   fmt.Sprintf("The number of permissions in the bundle is %d, but the number of permissions in the workspace is %d\n%s", len(b.Config().Permissions), len(objPermissions.AccessControlList), permissionDetails(objPermissions.AccessControlList, b.Config().Permissions)),
+		})
+		return diags
+	}
+
+	for _, p := range b.Config().Permissions {
+		level, err := permissions.GetWorkspaceObjectPermissionLevel(p.Level)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		found := false
+		for _, objPerm := range objPermissions.AccessControlList {
```

Comment:
If you turn `objPermissions` into a separate type then this logic can be extracted to a separate function, file, or even package. I feel it belongs more under `bundle/permissions` as a dedicated type for doing checks on bundle level permissions and actual permissions.